Add basic support for PDB

nico-abram opened the pull request:
See https://github.com/RazrFalcon/cargo-bloat/issues/17, I think this may be ready to merge as-is but it seems to be missing quite a few symbols. Not sure if merging initial support and possible improvements later is preferred, or waiting to see if we can get a better implementation.

Example output on this crate:

```
D:\dev\cargo-bloat [master ↑2]> cargo run ; cargo run --release -- --release
   Compiling cargo-bloat v0.10.1 (D:\dev\cargo-bloat)
    Finished dev [unoptimized + debuginfo] target(s) in 1.27s
     Running `target\debug\cargo-bloat.exe`
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
    Analyzing target\debug\cargo-bloat.exe

 File  .text    Size         Crate Name
 1.1%   1.4% 28.1KiB          json json::parser::Parser::parse
 0.5%   0.7% 13.2KiB          json json::util::print_dec::write
 0.5%   0.6% 11.8KiB  cargo_bloat? cargo_bloat::process_crate
 0.5%   0.6% 11.7KiB          json json::codegen::Generator::write_json
 0.5%   0.6% 11.6KiB          std? std::sys::windows::process::Command::spawn
 0.5%   0.6% 11.6KiB  cargo_bloat? cargo_bloat::parse_args
 0.4%   0.5% 10.6KiB  cargo_bloat? cargo_bloat::collect_pe_data
 0.3%   0.4%  8.0KiB           pdb <pdb::symbol::SymbolData as scroll::ctx::TryFromCtx>::try_from_ctx
 0.3%   0.4%  7.0KiB  regex_syntax <regex_syntax::hir::translate::TranslatorI as regex_syntax::ast::visitor::Visitor>::visit_class_set_item_post
 0.3%   0.3%  6.8KiB regex_syntax? regex_syntax::ast::parse::ParserI<ref_mut$<regex_syntax::ast::parse::Parser> >::parse_with_comments<ref_mut$<regex_syntax::ast::parse::Parser> >
 0.3%   0.3%  6.6KiB  regex_syntax <regex_syntax::hir::translate::TranslatorI as regex_syntax::ast::visitor::Visitor>::visit_post
 0.2%   0.3%  5.3KiB          std? alloc::collections::btree::map::BTreeMap::insert<std::sys::windows::process::EnvKey,std::ffi::os_str::OsString>
 0.2%   0.3%  5.3KiB         regex regex::exec::ExecBuilder::build
 0.2%   0.3%  5.3KiB           std core::num::flt2dec::strategy::dragon::format_shortest
 0.2%   0.3%  5.3KiB          std? alloc::collections::btree::map::BTreeMap::insert<std::sys::windows::process::EnvKey,enum$<core::option::Option<std::ffi::os_str::OsString>, 1, 1844674...
 0.2%   0.2%  4.7KiB  regex_syntax alloc::str::join_generic_copy
 0.2%   0.2%  4.5KiB      binfarce <binfarce::demangle::legacy::Demangle as core::fmt::Display>::fmt
 0.2%   0.2%  4.4KiB  cargo_bloat? cargo_bloat::main
 0.2%   0.2%  4.4KiB      binfarce binfarce::pe::Pe::symbols
 0.2%   0.2%  4.4KiB           std core::num::flt2dec::strategy::dragon::format_exact
69.2%  86.9%  1.7MiB               And 10270 smaller methods. Use -n N to show more.
79.6% 100.0%  1.9MiB               .text section size, the file size is 2.4MiB
   Compiling cargo-bloat v0.10.1 (D:\dev\cargo-bloat)
    Finished release [optimized] target(s) in 4.32s
     Running `target\release\cargo-bloat.exe --release`
    Finished release [optimized] target(s) in 0.02s
    Analyzing target\release\cargo-bloat.exe

 File  .text     Size Crate Name
 1.2%   1.6%  11.6KiB  std? std::sys::windows::process::Command::spawn
 0.6%   0.7%   5.3KiB  std? alloc::collections::btree::map::BTreeMap::insert<std::sys::windows::process::EnvKey,std::ffi::os_str::OsString>
 0.6%   0.7%   5.3KiB   std core::num::flt2dec::strategy::dragon::format_shortest
 0.6%   0.7%   5.3KiB  std? alloc::collections::btree::map::BTreeMap::insert<std::sys::windows::process::EnvKey,enum$<core::option::Option<std::ffi::os_str::OsString>, 1, 18446744073709...
 0.5%   0.6%   4.4KiB   std core::num::flt2dec::strategy::dragon::format_exact
 0.4%   0.5%   3.8KiB  std? rustc_demangle::v0::Printer::print_type
 0.3%   0.4%   3.3KiB   std rustc_demangle::demangle
 0.3%   0.4%   3.2KiB  std? std::sys::windows::process::Stdio::to_handle
 0.3%   0.4%   3.1KiB   std <rustc_demangle::legacy::Demangle as core::fmt::Display>::fmt
 0.3%   0.4%   3.0KiB   std std::process::Child::wait_with_output
 0.3%   0.4%   2.9KiB  std? rustc_demangle::v0::Printer::print_const
 0.3%   0.3%   2.6KiB  std? rustc_demangle::v0::Printer::print_path
 0.3%   0.3%   2.5KiB   std std::env::args_os
 0.2%   0.3%   2.3KiB   std core::num::flt2dec::strategy::grisu::format_shortest_opt
 0.2%   0.3%   2.1KiB  std? std::sys::windows::process::make_command_line::append_arg
 0.2%   0.2%   1.6KiB  std? std::backtrace_rs::print::BacktraceFrameFmt::print_raw_with_column
 0.2%   0.2%   1.6KiB  std? alloc::collections::btree::node::Handle::remove_leaf_kv<std::sys::windows::process::EnvKey,std::ffi::os_str::OsString,alloc::collections::btree::map::entry::...
 0.2%   0.2%   1.5KiB  std? std::panicking::default_hook
 0.2%   0.2%   1.5KiB   std <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
 0.2%   0.2%   1.5KiB  std? rustc_demangle::v0::impl$6::print_type::closure$0
12.6%  16.2% 121.3KiB       And 518 smaller methods. Use -n N to show more.
78.2% 100.0% 751.0KiB       .text section size, the file size is 960.5KiB
```

nico-abram commented:
Pushed a small change, it adds the unmangled symbols to the output so they at least show up. It's not great but some important ones were missing like `std::sys::windows::process::Command::spawn`. Updated the example output in the first message.

nico-abram commented:
Like I said in the issue comments, I tried llvm-pdbutil to see if I can find the missing unmangled symbols we need in dumps of the pdb but had no luck, I now also tried with the dia2dump tool that comes as a sample with windows' DIA SDK https://docs.microsoft.com/en-us/visualstudio/debugger/debug-interface-access/dia2dump-sample?view=vs-2022 with the `-all` parameter and did not find the unmangled symbols we need anywhere in the output.

I think we might need to ask rust compiler people where we may find them.

nico-abram commented:
I used a hex viewer ([ImHex](https://github.com/WerWolv/ImHex)) to look for strings in the pdb file I was using for testing and did not find any matches for one of the missing unmangled function names that is unmangled, so I think that pretty much confirms that they are indeed not present anywhere in the file.
![imagen](https://user-images.githubusercontent.com/24706838/146716454-2c9d1284-9826-40f1-a73a-0d63a5509b8d.png)

nico-abram commented:
Are unmangled names only needed to identify the crate name? Maybe we can just do a dumb hack and substring the beginning of the name until "::" to get it.

nico-abram commented:
I pushed a small change to assume V0 name mangling on the unmangled symbols, this seems to make the crate name finding work, but adds a "?" since it is "not_exact". Unless we get creative somehow or rustc improves it's PDB output with more mangled names, I'm not sure the current output can be improved. 

I don't think I'm gonna touch this PR any more unless there are review comments or similar suggestions.

RazrFalcon commented on src/main.rs:
```diff
@@ -860,17 +870,161 @@ fn collect_pe_data(path: &path::Path, data: &[u8]) -> Result<Data, Error> {
-        eprintln!("Warning: MSVC target is not supported.");
-        return Err(Error::UnsupportedFileFormat(path.to_owned()))
-    }
+        use pdb::FallibleIterator;
```
Move to a separate function. Like `collect_pdb_data`.

RazrFalcon commented on Cargo.toml:
```diff
@@ -16,6 +16,7 @@ multimap = { version = "0.8", default-features = false }
 pico-args = "0.4.2"
 term_size = "0.3.1"
 binfarce = "0.2.1"
+pdb="0.7.0"
```
Whitespaces.

RazrFalcon commented on src/main.rs:
```diff
@@ -860,17 +870,161 @@ fn collect_pe_data(path: &path::Path, data: &[u8]) -> Result<Data, Error> {
+
+        Ok(d)
+    } else {
+        let d = Data {
```
I think we can write:

```rust
Ok(Data {
    symbols,
    file_size: 0,
    text_size,
})
```

RazrFalcon commented on src/main.rs:
```diff
@@ -860,17 +870,161 @@ fn collect_pe_data(path: &path::Path, data: &[u8]) -> Result<Data, Error> {
+            };
+            path.with_file_name(file_name).with_extension("pdb")
+        };
+        let file = std::fs::File::open(&pdb_path).unwrap();
```
Handle error. No unwrap.

RazrFalcon commented on src/main.rs:
```diff
@@ -860,17 +870,161 @@ fn collect_pe_data(path: &path::Path, data: &[u8]) -> Result<Data, Error> {
+        let symbol_table = pdb.global_symbols()?;
+        let address_map = pdb.address_map()?;
+
+        let mut out_symbols = vec![];
```
Prefer `Vec::new()`

RazrFalcon commented on src/main.rs:
```diff
@@ -860,17 +870,161 @@ fn collect_pe_data(path: &path::Path, data: &[u8]) -> Result<Data, Error> {
+                        public_symbols.push((data.offset, data.name.to_string().into_owned()));
+                    }
+                    if data.name.to_string().contains("try_small_punycode_decode") {
+                        dbg!(&data);
```
Remove

RazrFalcon commented on src/main.rs:
```diff
@@ -860,17 +870,161 @@ fn collect_pe_data(path: &path::Path, data: &[u8]) -> Result<Data, Error> {
+                        public_symbols.push((data.offset, data.name.to_string().into_owned()));
+                    }
+                    if data.name.to_string().contains("try_small_punycode_decode") {
+                        dbg!(&data);
```
Remove.

RazrFalcon commented:
Can you show output with `--crates`?

nico-abram commented:
Addressed the review comments. I can squash the commit if needed.

Output from `--crates` on cargo-bloat:

```
D:\dev\cargo-bloat [master ≡ +0 ~2 -0 ~]> cargo bloat --crates
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
    Analyzing target\debug\cargo-bloat.exe

 File  .text     Size Crate
19.3%  24.2% 479.4KiB std
17.7%  22.2% 438.7KiB regex_syntax
 9.5%  12.0% 237.0KiB regex
 8.7%  10.9% 215.5KiB binfarce
 8.6%  10.9% 214.6KiB pdb
 5.6%   7.1% 139.6KiB json
 3.9%   4.9%  96.1KiB cargo_bloat
 1.6%   2.0%  40.0KiB pico_args
 0.5%   0.6%  11.6KiB [Unknown]
 0.3%   0.4%   8.2KiB memmap2
 0.1%   0.2%   3.4KiB uuid
 0.1%   0.2%   3.3KiB scroll
 0.0%   0.1%   1.2KiB multimap
 0.0%   0.0%     526B term_size
79.6% 100.0%   1.9MiB .text section size, the file size is 2.4MiB
```

RazrFalcon commented:
>Maybe we can just do a dumb hack and substring the beginning of the name until "::" to get it. 

I haven't touched the code in years, so I don't remember why it's needed. I assume because of `<regex_syntax::hir::...`.

Based on `--crates` output, less then 1% was not resolved. This is to be expected. Current parses are also fail to resolve all the data.

RazrFalcon commented:
Thanks for your work. I really don't have time working on it. And I don't use Windows.
I will merge your changes as is, then will try to tests it myself this week and will probably publish a new version closer to weekends.

RazrFalcon commented:
@nico-abram Hi! Is it me, or release build is broken? I see only `std` functions there. Debug is fine.
It seems like it doesn't find any info on non-std functions.
It even visible in your output.

nico-abram commented:
You're right. Adding this to Cargo.toml fixes it:

```rs

[profile.release]
debug=true
```

output:
```
D:\dev\cargo-bloat [master ≡ +1 ~2 -0 !]> cargo run --release -- --release --crates
    Finished release [optimized + debuginfo] target(s) in 0.02s
     Running `target\release\cargo-bloat.exe --release --crates`
    Finished release [optimized + debuginfo] target(s) in 0.02s
    Analyzing target\release\cargo-bloat.exe

 File  .text     Size Crate
28.3%  36.1% 270.5KiB std
16.1%  20.6% 154.4KiB regex_syntax
10.9%  13.9% 104.2KiB cargo_bloat
 7.8%  10.0%  74.5KiB regex
 5.2%   6.7%  50.1KiB binfarce
 3.9%   5.0%  37.3KiB pdb
 2.7%   3.4%  25.4KiB json
 0.5%   0.6%   4.9KiB pico_args
 0.3%   0.4%   2.7KiB [Unknown]
 0.2%   0.2%   1.8KiB memmap2
 0.1%   0.1%     537B scroll
 0.0%   0.1%     467B multimap
 0.0%   0.0%     132B term_size
78.2% 100.0% 748.5KiB .text section size, the file size is 957.0KiB

Note: numbers above are a result of guesswork. They are not 100% correct and never will be.
```

I feel like this is to be expected, as release has no debug symbols by default.

RazrFalcon commented:
But the whole point of cargo-bloat is to check release builds. Debug output is pointless, because optimizer will throw all of this away.
All other formats still have all the information needed by cargo-bloat. So we have to either figure how to get it in a release build or remove PDB support.

Yes, we can add a warning about PDB + release build, but it seems like a dirty hack.

nico-abram commented:
I tried comparing against this sizer tool and it seems it is also unable to get symbols unless they are present in the PDB https://github.com/aras-p/sizer/

I also checked using x64dbg that no symbols other than std seem to be present in the pdb/PE unless debug info is enabled for release builds by using it's symbols tab on the executable.

Note: Just in case, I wanna clarify that all `debug=true` does is enable debug info, it has no effect whatsoever on code generation AFAIK, it just makes it so rust adds the symbols to the PDB. See the cargo book for documentation on it https://doc.rust-lang.org/cargo/reference/profiles.html#debug
![imagen](https://user-images.githubusercontent.com/24706838/147695264-9c5a0062-5143-4fe3-a638-a9256c251bd4.png)

RazrFalcon commented:
Hmm... I guess I would have to add a warning then...

RazrFalcon commented:
Or maybe we could pass it automatically?

nico-abram commented:
I asked in the rust zulip (https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/rust.20MSVC.20target.20fn.20debug.20symbols.20with.20debug.3Dfalse)

bjorn mentioned setting the `CARGO_PROFILE_RELEASE_DEBUG=true` environment variable would force it, but that forces a full recompile of the project when running cargo bloat, and then another recompile when the user tries to compile their code again via the normal cargo build --release .

I feel like it is also surprising behaviour. My understanding is there are reasons release builds have debug symbols off by default, for example shipping propietary software without symbols. Randomly making it so symbols *are* compiled in a hidden way by silently setting a flag sounds potentially bad/undesirable to me. I guess since the PDB is a separate file maybe it's not that big a deal, since they would have to also ship that file, but I don't know if rustc embeds any other undesirable information in the PE binary itself when debug=true is enabled

nico-abram commented:
Laurențiu on zulip suggested using a different target directory in combination with that environment flag, I think that fixes my concern about confusing the cargo bloat with the actual release build and it also helps a bit with the recompile problem in some cases.

RazrFalcon commented:
>I feel like it is also surprising behaviour. 

Well, this is _the default_ behaviour for all other targets. And no one should use local builds after running 3rd-party tools. This is not our problem.

Having a rebuild is not a problem in my eyes. The only problem left is how to figure out that our current target is MSVC. I don't want `CARGO_PROFILE_RELEASE_DEBUG` to affect all targets.

nico-abram commented:
I believe we can use `rustc --print cfg` (From https://stackoverflow.com/questions/52996949/how-can-i-find-the-current-rust-compilers-default-llvm-target-triple?rq=1)
```
D:\dev\cargo-bloat [master ≡ +1 ~2 -0 !]> rustc --print cfg
debug_assertions
target_arch="x86_64"
target_endian="little"
target_env="msvc"
target_family="windows"
target_feature="fxsr"
target_feature="sse"
target_feature="sse2"
target_os="windows"
target_pointer_width="64"
target_vendor="pc"
windows
```

If target_os="windows" and target_env="msvc", I think we can be pretty confident the target is msvc. Then we just need to also check for overrides of these via cargo args (Like --target some-other-triplet)

RazrFalcon commented:
Nice! Thanks for an investigation. I will look into it tomorrow.
